feature: Add optional right-swipe-to-undo with dynamic background

Politanskyi opened the pull request:
This pull request introduces a new optional feature to provide a more intuitive and visually consistent user experience when implementing custom swipe logic, such as "swipe right to undo".

The Problem:

Currently, the card revealed in the background during a swipe is always the next card in the stack (index + 1). This behavior is hardcoded and cannot be changed. This creates a visual disconnect for users when an application configures a right-swipe to be an "undo" action, as the user expects to see the previous card being revealed.

The Solution:

This feature is enabled by a new boolean parameter on the CardSwiper widget: showBackCardOnUndo (defaults to false to ensure backward compatibility).

When showBackCardOnUndo is set to true:

The swiper detects the horizontal swipe direction as the user drags the card.

If swiping right, the previous card from the stack is shown in the background.

If swiping left, the default behavior is maintained, and the next card is shown.

This allows developers to create a more polished and logical interface where the visual feedback perfectly matches the action performed.

Politanskyi committed:
feature: Add optional right-swipe-to-undo with dynamic background

This commit introduces a new optional feature that enhances the swiper's flexibility and provides a more intuitive user experience for looping carousels.

Currently, the undo action fails when triggered on the first card (index 0) in a looping swiper, as there is no previous state in the history. Additionally, the background card is always the next card in the stack, which can be visually confusing when a right swipe is configured to act as an "undo."

This feature, controlled by the new showBackCardOnUndo boolean property, addresses these issues by:

Enabling True Looping Undo: When showBackCardOnUndo is true and a user swipes right on the first card, the swiper now correctly loops to the last card in the stack with a proper undo animation.

Providing Dynamic Backgrounds: During a right swipe, the swiper will now display the previous card in the background, creating a more logical and visually consistent interaction. During a left swipe, it retains the default behavior of showing the next card.

Ensuring Backward Compatibility: The feature is disabled by default (showBackCardOnUndo: false), so there are no breaking changes for existing users of the package.

This enhancement makes the CardSwiper more powerful for developers looking to implement advanced and non-linear navigation patterns.

Politanskyi committed:
ci: normalize line endings for Dart files

Politanskyi committed:
ci: remove line endings normalization step from code check

Politanskyi committed:
feature: remove unsafe_html lint rule from analysis options

Politanskyi committed:
fix: update BoxShadow color to use withValues for better opacity control

Cyrille37 commented:
Hi @Politanskyi 

I think changing the next card could produce a very strange result for the user like I show in comment on issue [#45](https://github.com/ricardodalarme/flutter_card_swiper/issues/45#issuecomment-3187189732). When starting to swipe you already see the back card while you can change the direction.

Politanskyi committed:
Merge branch 'main' into feature/optional-undo-swipe

Politanskyi committed:
feature: update minimum iOS version to 13.0 and add undo swipe threshold to CardSwiper

Politanskyi committed:
feature: remove commented-out onPanUpdate logic from CardSwiper

Politanskyi commented:
Hi @ricardodalarme and @Cyrille37  did you have time to look at pull request?

Politanskyi commented:
> @ricardodalarme

@ricardodalarme perhaps you have time to look at the latest  PR?

Politanskyi commented:
@ricardodalarme perhaps you have time to look at the latest PR?

ricardodalarme commented on .github/workflows/code-check.yml:
```diff
@@ -45,7 +45,7 @@ jobs:
 
       - name: Run Flutter Format
-        run: dart format --set-exit-if-changed .
+        run: dart format .
```
pls lets keep `--set-exit-if-changed` because we want to formatter to tell when there are any format issues in the code

ricardodalarme reviewed (dismissed):
LGTM just pls remove the --set-exit-if-changed pls

ricardodalarme commented:
If you could also add some test I'd be very grateful

Politanskyi committed:
✨ Add undo swipe functionality to CardSwiper

- Implemented logic to show the previous card when swiped beyond the undo threshold.
- Ensured valid index handling for card navigation.
- Added tests to verify behavior of showBackCardOnUndo feature.

Politanskyi committed:
♻️ Refactor code formatting command in CI configuration

- Changed the Flutter format command to remove the exit condition.
- This allows for better integration with CI workflows without failing the build on formatting changes.

Politanskyi committed:
♻️ Refactor CardSwiper test cases to remove unused parameters

- Removed `numberOfCardsDisplayed` and `undoSwipeThreshold` from test cases.
- Simplified test setup for better clarity and maintainability.

Politanskyi commented:
> If you could also add some test I'd be very grateful

all done waiting for your check @ricardodalarme

ricardodalarme reviewed (approved):
LGTM thanks for your contribution and patience 🙂

txufiknr commented:
What about vertical direction (top/bottom) `undoDirection`? Should have that, aligning with `allowedSwipeDirection`
